src/optimagic/optimization/optimize.py

```diff
@@ -605,6 +557,29 @@ def _optimize(problem: OptimizationProblem) -> OptimizeResult:
     else:
         logger = None
 
+    # ==================================================================================
+    # Strict checking if bounds are required and infinite values in bounds
+    # ==================================================================================
+    if problem.algorithm.algo_info.supports_bounds:
+        if problem.algorithm.algo_info.needs_bounds:
+            if (
+                internal_params.lower_bounds is None
+                or internal_params.upper_bounds is None
+            ):
+                raise IncompleteBoundsError(
+                    f"Algorithm {problem.algorithm.name} needs finite bounds "
+                    f"for all parameters. "
+                )
+        if not problem.algorithm.algo_info.supports_infinite_bounds:
+            if (
+                np.isinf(internal_params.lower_bounds).any()
+                or np.isinf(internal_params.upper_bounds).any()
+            ):
+                raise IncompleteBoundsError(
```

Comment:
Conceptually, this code will only be correct after #603 is implemented. 

Right now, if the user specifies no bounds, the lower and upper bounds of the internal_params object will not be None but arrays full of `-np.inf` and `np.inf`. 

To make it correct we would have to extend the condition in line 566 and 567 to: 

```python
internal_params.lower_bounds is None
or internal_params.upper_bounds is None
or np.isinf(internal_params.lower_bounds).all()
or np.isinf(internal_params.upper_bounds).all()
```
Can you do this? Other than that the PR is great and ready to be merged.